src/main/scala/nl/biopet/tools/gtftorefflat/GtftoRefflat.scala

```diff
@@ -1,17 +1,186 @@
+              feature.minPosition,
+              feature.maxPosition,
+              coding._1,
+              coding._2,
+              exonBuffer.remove(geneId, transcriptId).getOrElse(Nil))
+            if (genesBuffer.contains(geneId)) {
+              val oldGene = genesBuffer(geneId)
+              genesBuffer(geneId) = oldGene.copy(transcripts = transcript :: oldGene.transcripts)
+            } else
+              transcriptsBuffer(geneId) = transcript :: transcriptsBuffer.getOrElse(geneId, Nil)
+          case "exon" =>
+            val exon = Exon(feature.minPosition, feature.maxPosition)
+            genesBuffer.get(geneId).flatMap(_.transcripts.find(_.name == transcriptId)) match {
+              case Some(transcriptOld) =>
+                val geneOld = genesBuffer(geneId)
+                genesBuffer(geneId) = geneOld.copy(
+                  transcripts = transcriptOld
+                    .copy(exons = exon :: transcriptOld.exons) :: geneOld.transcripts.filter(
+                    _.name != transcriptOld.name))
+              case _ =>
+                transcriptsBuffer.get(geneId).flatMap(_.find(_.name == transcriptId)) match {
+                  case Some(transcriptOld) =>
+                    transcriptsBuffer(geneId) = transcriptOld.copy(
+                      exons = exon :: transcriptOld.exons) ::
+                      transcriptsBuffer(geneId).filter(_.name != transcriptId)
+                  case _ =>
+                    exonBuffer((geneId, transcriptId)) = exon :: exonBuffer.getOrElse(
+                      (geneId, transcriptId),
+                      Nil)
+                }
+            }
+          case "stop_codon" if !feature.strand.contains(false) =>
+            updateCodingRegion(geneId, transcriptId, end = Some(feature.end))
+          case "stop_codon" =>
+            updateCodingRegion(geneId, transcriptId, start = Some(feature.start - 1))
+          case "start_codon" if !feature.strand.contains(false) =>
+            updateCodingRegion(geneId, transcriptId, start = Some(feature.start - 1))
+          case "start_codon" => updateCodingRegion(geneId, transcriptId, end = Some(feature.end))
+          case "CDS" =>
+            updateCodingRegion(geneId,
+              transcriptId,
+              start = Some(feature.start - 1),
+              end = Some(feature.end))
+          case _ =>
+        }
+      }
+
+    featureBuffer.foreach { case (k, c) => logger.info(s"$k\t$c") }
+
+    val writer = new PrintWriter(refflatFile)
+
+    for {
+      (_, genes) <- genesBuffer.values.toList.groupBy(_.contig).toList.sortBy(_._1)
+      gene <- genes.sortBy(_.start)
+      transcript <- gene.transcripts
+    } {
+      val exons = transcript.exons.sortBy(_.start)
+      val values = List(
+        gene.name,
+        transcript.name,
+        gene.contig,
+        if (gene.strand) "+" else "-",
+        (transcript.transcriptionStart - 1).toString, //TODO: check if this is correct
```

Comment:
Is this correct?

Reply:
yes, did some extended testing when creating this tool on this. Some formats are 0 based, some are 1 based. Even worse, sometimes start is 1 based and end is 0 based